api/wasm/v0/cpp/proxy_wasm_externs.h

```diff
@@ -0,0 +1,126 @@
+ */
+extern "C" uint32_t proxy_validate_configuration(uint32_t root_context_id,
+                                                 uint32_t configuration_size);
+/**
+ * Called when a plugin loads or when plugin configuration changes dynamically.
+ * @param root_context_id is an identifier for one or more related plugins.
+ * @param plugin_configuration_size is the size of any configuration available via
+ * proxy_get_configuration().
+ * @return non-zero on success and zero on failure (e.g. bad configuration).
+ */
+extern "C" uint32_t proxy_on_configure(uint32_t root_context_id,
+                                       uint32_t plugin_configuration_size);
+
+// Stream calls.
+
+/**
+ * Called when a request, stream or other ephemeral context is created.
+ * @param context_id is an identifier the ephemeral context.
+ * @param configuration_size is the size of any configuration available via
+ * proxy_get_configuration().
+ */
+extern "C" void proxy_on_context_create(uint32_t context_id, uint32_t root_context_id);
+
+// Stream and Non-stream calls.
+
+/**
+ * For stream contexts, called when the stream has completed. Note: if applicable proxy_on_log() is
+ * called after proxy_on_done() and before proxy_on_delete(). For root contexts, proxy_on_done() is
+ * called when the VM is going to shutdown.
+ * @param context_id is an identifier the context.
+ * @return non-zero to indicate that this context is done. Stream contexts must return non-zero.
+ * Root contexts may return zero to defer the VM shutdown and the proxy_on_delete call until after a
+ * future proxy_done() call by the root context.
+ */
+extern "C" uint32_t proxy_on_done(uint32_t context_id);
+/**
+ * Called when the context is being deleted and will no longer receive any more calls.
+ * @param context_id is an identifier the context.
+ */
+extern "C" void proxy_on_delete(uint32_t context_id);
+
+//
+// ABI calls from the VM to the host.
+//
+
+// Configuration and Status
+
+/**
+ * Called from the VM to get any configuration. Valid only when in proxy_on_start() (where it will
+ * return a VM configuration), proxy_on_configure() (where it will return a plugin configuration) or
+ * in proxy_validate_configuration() (where it will return a VM configuration before
+ * proxy_on_start() has been called and a plugin configuration after).
+ * @param configuration_ptr a pointer to a location which will be filled with either nullptr (if no
+ * configuration is available) or a pointer to a malloc()ed block containing the configuration
+ * bytes.
+ * @param configuration_size a pointer to a location containing the size (or zero) of any returned
+ * configuration byte block.
+ * @return a WasmResult: OK, InvalidMemoryAccess. Note: if OK is returned  *configuration_ptr may
+ * be nullptr.
+ */
+extern "C" WasmResult proxy_get_configuration(const char** configuration_ptr,
+                                              size_t* configuration_size);
```

Comment:
I get that there are language SDKs on top of this ABI, but I still find it bizarre that you have this double call API. Why isn't the configuration just passed in the relevant calls again? I cannot see any good reason to have this extra complexity. Can't the host just pass a constant byte buffer and if the VM code wants to copy it it can do that?

Reply:
The Wasm VM can only access memory in the VM.  Consequently, all data must be passed as integral call parameters or by malloc()ing data in the VM.  For consistency and to enable diverse Wasm languages (e.g. languages with GC), we have a single mechanism for allocating memory in the VM and we require that all memory allocations be explicitly requested by calls from the VM.  This was in direct response to leaks from our early users.  There is additional discussion in the Architecture document.

Reply:
OK this makes sense. Can you put a high level note at the top of this file with roughly your response? I think this question is going to keep coming up as it's not clear to the non-WASM initiated.

Reply:
Added header comments.